### ISSUE TITLE
Mandate pull request merge only after branch-deploy workflow succeeds

### ISSUE THREAD

**Issue by bzurkowski** [github/branch-deploy#160]
### Details

I am evaluating the branch-deploy action as part of a Terraform-infrastructure deployment workflow.

My process mandates that a pull request can only be merged post the successful application of all infrastructure changes to target environments.

The obvious solution would be to configure jobs of the branch-deploy workflow, e.g., the Terraform `apply` job, as status checks in branch protection rules. However, since the branch-deploy workflow runs in the context of the main branch - the consequence of the `on: issue_comment` trigger - status checks are never reported to the feature-branch and thus never succeed.

I am wondering if there is a solution to this within the scope of the IssueOps approach, or if a "hacky" approach is necessary, e.g., manual status reporting via the GitHub API? I would appreciate any insights if you have encountered a similar process or use case. Thanks!

**Comment by bzurkowski** [comment]
@GrantBirki Thank you for prompt response and step-by-step instructions 🥇 

This is indeed a brilliant strategy that utilizes the native capabilities of Github. I am quite fond of it! It is a solution that should be effective for a vast number of action users.

However, I handle infrastructure code in a mono-repo, which necessitates the deployment of numerous projects across various environments from a single repository. Furthermore, depending on the scope of implemented changes, I need to deploy these projects selectively.

With the strategy you proposed, I would need to require deployment in each Github environment (corresponding to respective projects environments). That would force me to deploy all projects across every environment on each pull request.

Do you know if the proposed approach can be adjusted to the mono-repo characteristics or if there is any alternative approach?

**Comment by GrantBirki** [comment]
@bzurkowski What you are describing is a know issue and I have personally run into before. From my experience on personal projects and professionally, I have always run into hiccups with monorepos. There are certainly advantages and disadvantages to every project structure and a disadvantage here is this very limitation you described.

With GitHub branch protections and this Action, we are kinda stuck. I personally have projects that are exactly the nature that you described and ran into the same issue.

What we really need is: "Somehow detect when changes are made to environment ABC and then require a deployment to that environment"

And then in the case of multiple changes within the monorepo: "Changes were made to ABC and XYZ so we require deployments to both environments before merging"

Unfortunately, I am unsure how to actually implement this in a meaningful way. I am totally open to suggestions and would be happy to write the code for it if a suggestion surfaces that would work!

**Comment by bzurkowski** [comment]
@GrantBirki

My idea to tackle this issue is to leverage noop and non-noop deployments as hooks for managing `apply` check runs.

The noop deployment (as a deployment intent) could be used to register `apply` checks. In turn, the non-noop deployment could be used to conclude corresponding checks.

[Github Checks API](uns?apiVersion=2022-11-28) could be used to implement these operations.

Here is a plan for how the above strategy could be implemented:

1. On noop-deployment (Terraform plan), use Github API to [queue](https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#create-a-check-run) a PR status check for plan application, using a naming convention, e.g., `<environment>-apply`. As a check `ref` use output provided by the `branch-deploy` trigger.
2. On deployment (Terraform apply), use Github API to [conclude](https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#update-a-check-run) corresponding status check. In order to obtain check ID (for update operation), [filter status checks](https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#list-check-runs-for-a-git-reference) by `check_name` according to the assumed naming convention.
3. Add an additional (helper) job at the end of the deployment workflow, e.g., `ensure-applied`. Filter `apply` check runs and verify whether they all concluded with success.

Please, let me know what you think and if you see a sense in employing a similar flow in the action 😉

**Comment by bzurkowski** [comment]
@GrantBirki I tried to implement a PoC of the above strategy. Check registration on noop deployment works fine 🚀  However, I encountered an issue at the apply stage. The `branch-deploy` trigger blocks the deployment because registered check runs make the commit stuck in a `PENDING` status. I get error:

```
Error: ### ⚠️ Cannot proceed with deployment

- reviewDecision: `null`
- commitStatus: `PENDING`
```

The `required_contexts` parameter on the actions does not seem to help at all.

**Comment by GrantBirki** [comment]
I really like the idea of registering "checks" based on deployments here. Internally, I have been asked before if it is possible to "require" a `noop` deploy before a `full` deploy and my answer until today has been _"probably not"_. However, with what you have described here, it looks like that might actually be a possibility. I'll need to do some thinking though on how this could actually be implemented as the `branch-deploy` Action internally does a lot of things which check suites and if they are in a pending state it will reject the request. So there needs to either be some modifications to that code, or for check suites to be added in a way that don't conflicts with a repo's existing check suites around testings, linting, code scanning, etc.

I'm not sure what my availability look like for actually working on this in the near future. Is this something that you are interested in opening a pull request for?

**Comment by bzurkowski** [comment]
I am afraid my availability does not allow me to spend enough time to get into the code base and propose an implementation.

However, I discovered that Github status checks can be created with a `completed` `status`, while the `conclusion` can be set to `neutral`. Perhaps, a combination of these could help us achieve what we need. I will check that next week.

**Comment by GrantBirki** [comment]
It doesn't look like there is any path forward at this time. I'll provide a brief TL;DR of what was discussed in this issue below before closing this out:

**TL;DR**; If you are using the branch-deploy Action in a monorepo project that has multiple "things" to deploy and multiple "environments" to deploy to, you may run into issues using branch protection rules to properly gate your deployments. You might require that `production` requires a deployment before merging, but using native GitHub features, you won't be able to chose which part of your monorepo requires that deployment. This isn't great and there isn't really any path being worked on at this time that would resolve this.

For all other users that are not using a monorepo structure with multiple environments, you will likely not be impacted by this issue in anyway.

---

I am going to close this issue out *for now*. If anyone has an idea or would like to chat about a potential solution for this, feel free to re-open this issue and we can discuss.
